### PR TITLE
Use two lines for printer driver info in Add Dialog

### PR DESCRIPTION
**Pull request by jeremypw**
Modifies #158 

An attempt to implement @danrabbit 's design for the printer driver info, using a listbox.

Constructing the listbox synchronously resulted in the interface freezing for a noticeable time for e.g. HP as there are nearly 5000 entries in that case. This was improved by making the construction asynchronous and cancellable.

### CODE REVIEW

**Comment by jeremypw** [comment]
AddDialogTwoLineInfo

Using a single label + markup.  Are there style classes that would be better?

File: src/Dialogs/AddDialog.vala [hunk 1]
```diff
@@ -643,4 +640,32 @@ public class Printers.AddDialog : Granite.Dialog {
...
+    public class DriverRow : Gtk.ListBoxRow {
+        public DeviceDriver driver { get; construct; }
+        public DriverRow (DeviceDriver driver) {
+            Object (driver: driver);
+        }
+
+        construct {
+            var model = driver.ppd_make_and_model;
+            model = model.replace ("(recommended)", _("(recommended)"));
+            var model_markup = ("<span size='large' weight='bold'>%s</span>").printf (model);
```
Comment (tintou) [inline src/Dialogs/AddDialog.vala:653]:
```suggestion
            var model_markup = GLib.Markup.printf_escaped ("<span size='large' weight='bold'>%s</span>", model);
```

File: src/Dialogs/AddDialog.vala [hunk 2]
```diff
@@ -643,4 +640,32 @@ public class Printers.AddDialog : Granite.Dialog {
...
+
+            var detail_markup = ("<span size='medium' weight='normal'>%s — %s</span>").printf (
```
Comment (tintou) [inline src/Dialogs/AddDialog.vala:655]:
```suggestion
            var detail_markup = GLib.Markup.printf_escaped ("<span size='medium' weight='normal'>%s — %s</span>",
```

File: src/Dialogs/AddDialog.vala [hunk 3]
```diff
@@ -643,4 +640,32 @@ public class Printers.AddDialog : Granite.Dialog {
...
+                driver.ppd_natural_language, driver.ppd_name
+            );
+
+            var label = new Gtk.Label ("") {
+                halign = Gtk.Align.START,
+                margin = 6,
+                use_markup = true
+            };
+
+            label.set_markup (("%s\n%s").printf (model_markup, detail_markup));
```
Comment (tintou) [inline src/Dialogs/AddDialog.vala:665]:
```suggestion
            label.set_markup ("%s\n%s".printf (model_markup, detail_markup));
```

**Comment by andrewdbate** [comment]
IMHO the font of the bold text in the list looks a little too large to me. Shouldn't the font of everything in the list be the same size as the manufacturer list on the left? It just looks a bit huge.

**Comment by jeremypw** [comment]
I first used "medium" and "small" sizes for the two lines but found that the "small" info was hard to read so I erred on the side of legibility.

**Comment by jeremypw** [comment]
Does anyone know whether the different ppd URLs for the apparently duplicate make/model entries really point to different files?

**Comment by jeremypw** [comment]
According to HP docs, ppd files can be dynamic or static but I do not understand HPLIP at present.

**Comment by jeremypw** [comment]
I don't know how much the average user would make of these ppd URLs ...

**Comment by andrewdbate** [comment]
I agree that the average user (or even an expert user who just isn't an expert in Linux printer drivers) probably won't understand these PPD URIs. It might not even be obvious that it somehow uniquely identifies the driver. They might look at the URI and say "what the hell this that?".

I'm no expert here, but according to https://www.cups.org/doc/admin.html, under section "Printer Drivers and PPDs" it says that the command `lpinfo -m` produces an output such as:
```
drv:///sample.drv/dymo.ppd Dymo Label Printer
drv:///sample.drv/epson9.ppd Epson 9-Pin Series
drv:///sample.drv/epson24.ppd Epson 24-Pin Series
drv:///sample.drv/generpcl.ppd Generic PCL Laser Printer
drv:///sample.drv/generic.ppd Generic PostScript Printer
```
where each line contains _the driver name followed by its description_.

My understanding is therefore that the URI is the driver name. Is this correct?

If so, should we make it clear that these URIs are the driver names? Then it should be clear that this string uniquely identifies the driver.

**Comment by jeremypw** [comment]
It seems HP ppds have various weird URLS, even if the basename is the same. Not all printers are consistently associated with a `drv:///` URL or a `hplip:` URL unfortunately. I guess we will just have to show them all until someone works out which, if any, can be ignored :shrug:

**Comment by jeremypw** [comment]
@tintou Are you happy to merge into your PR?

**Review by danrabbit — APPROVED** [review]

**Pull request merged**